added dockerized test setup

schmunk42 opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | yes
| Tests pass?   | it's about testing
| Fixed issues  | relates https://github.com/yiisoft/yii2/issues/13047

essential

- [x] base-image (rela
- [x] mssql stack
- [x] cubrid stack
- [ ] host-volumes (yes/no?)
- [ ] reports

optional

- [x] additonal tools/dev-image(?)
- [x] isolation in GitLab CI

related: https://github.com/Codeception/Codeception/pull/3739

schmunk42 committed:
updated debug settings (+56 squashed commits)
Squashed commits:
[c42f30c] updated base image, added docs & local test script
[88f0c40] debug
[6220c94] fixed network isolation
[c63c7c3] test mssql only on test/mssql branch
[74efc78] fixed isolation in after_script
[02b895b] updated test setup
[3335f39] updated retry
[bd123b2] updated service checks
[45e4c90] updated build
[d54da7a] updated after_script
[5a4c726] :factory: wait for mysql
[08db878] fixed typo
[ea53c1e] updated build stages
[9807ce3] fixed typos
[cf9f64e] fixed mssql testing
[08001d6] added db create for mssql
[62f6b65] run travis (gitlab simulation) only in travis branch
[cf63da4] streamlined build
[76808ac] updated test jobs
[18d79b5] fixed test error
[7b2bce6] updated build & composer.lock
[244623a] updated build
[86bd71b] fixed cleanup
[86ab2e8] fixed cleanup
[091d4b8] fixed tests
[2d315b5] fixed build config
[2913644] fixed project names
[f53b823] refactored build config
[5a791fb] refactored docker db-tests
[b4479b0] revert
[a975fa5] updated gitlab build
[4e4e5e4] updated mssql setup
[d6ff03b] added sleep workaround
[578b102] removed host volumes in test
[928f50b] fixed path
[967ab10] updated tests
[520f317] bootstrap cubrid
[5f245e1] :factory: fixed cubrid tests
[940dbbc] :factory: pinned cubrid version 9.3.6.0002
[8d5ea69] :memo: dockerized test commands
[9954b54] updated cubrid
[fb3afac] updated docs
[3f63ced] updated isolation
[bdc3c83] :factory: build cubrid
[b777911] :factory: disabled mssql, updated cubrid
[36534e7] added travis-a-like stage
[6546f02] updated GitLab build
[c046096] updated test setup
- added mssql and cubrid stack
[5a0e635] disabled host-volume
[75cf342] fixed test
[8de0794] fixed testing
[465d27a] added stages
[63aa950] added script
[68eecef] fixed typo
[00e4b88] updated Docker build
[f9072cc] added dockerized test setup

samdark commented on tests/docker-compose.yml:
```diff
@@ -0,0 +1,32 @@
+      - POSTGRES_PASSWORD=postgres
+
+  #sqlsrv:
+  #  image: microsoft/mssql-server-linux
```
Are you going to enable it as well?

schmunk42 replied:
Yes, I tried, but you need to install some additional packages - just did not work right away. Let me know if you've some input here, I've never worked with MSSQL.

samdark replied:
I have no experience with dockerizing it or even running it under Linux.

schmunk42 commented:
## TL;dr

So, theoretically this is a complete **setup**. You can view an example GitHub build [here](https://git.hrzg.de/schmunk42/yii2/pipelines/17792) - they're not green yet.

I've split the setup into 3 stacks: default, mssql and cubrid. We can add more later. There are 4 test jobs, which are run in parallel. Whenever possible I've used pre-built images.

Currently there are different base-images, I intentionally left it in this state, because it shows a few issues we should sort out, before continuing.

### Choosing a base image

There are several options how to create a Docker image for testing. You can build from different base-images, like `php`, `composer` or `codeception/codeception` or `debian`, `ubuntu`, ....
Currently there's no official Yii image, another option.

Basically you need to install the required PHP-extensions for Yii (that's why some tests fail, because this is just a vanilla PHP) and for some suites also additional extensions.

I'd say it's not really possible to build a Yii image with all extensions (pdo_mssql, pdo_cubrid, xcache etc...) just because some are not available for `debian`, etc... So I'd just choose what works, i.e. `codemix` base-image, custom build for cubrid, pre-build for mssql.

In tests, there are dependencies from i.e. log suite to db, which make tests fail which would actually run on a vanilla PHP. I thought about also putting them into group `db`, WDYT?
These components might also be reviewed for decoupling.

Depending on how you wanna build the image you also need composer and some other tools, like npm or yarn, i.e. for testing asset bundling via webpack. But that's a separate topic.

Tests do not pass on Travis now, this is just because of changed DB credentials and hosts (should be easy to fix)

---

While I could easily write 5x times more :) I'd like stop here and also CC: @mikehaertl and @motin - your experience and feedback about Docker, PHP & Yii is very welcome.

Would be really nice if you could give it a try and just report if the things start to work.

motin commented:
Love it. Every open source project should offer a way of running database-dependent tests without having to follow extensive complicated instructions on how to set up databases and credentials. @schmunk42, please supply a single shell-script that will run all tests with all database engines in sequence and I'll verify that it works over here :)

schmunk42 commented:
This should be ready to try out locally.
Let me know if you've questions...

schmunk42 committed:
updated docs

schmunk42 committed:
fixed cleanup

schmunk42 committed:
fixed network setup

schmunk42 committed:
fixed tuple

schmunk42 committed:
Merge commit '3f66fb6cdfd498a119bfb34562549d07ff2ff9e0' into feature/docker-testing

schmunk42 committed:
run cubrid only on branch tests/cubrid

schmunk42 committed:
updated build
- added git files to dockerignore
- updated local testing
- updated build on branches

schmunk42 commented:
Example build pipelines
- https://git.hrzg.de/schmunk42/yii2/pipelines/18231 (all, except oci & mssql)
- https://git.hrzg.de/schmunk42/yii2/builds/72942 (like travis but twice as fast)

schmunk42 committed:
Merge commit 'd4d6c1dabf555e6f14a668df6fcc515bceef2830' into feature/docker-testing

schmunk42 committed:
updated build branches

Commit:
Merge branch 'feature/docker-testing' into 'tests/mysql'

updated build branches

See merge request !1

schmunk42 commented on tests/data/config.php:
```diff
@@ -16,29 +16,29 @@
     'databases' => [
         'cubrid' => [
-            'dsn' => 'cubrid:dbname=demodb;host=localhost;port=33000',
+            'dsn' => 'cubrid:dbname=demodb;host=cubrid;port=33000',
```
@cebe @samdark How should we handle this? Using a separate file for Docker tests. It's theoretically possible to fake this with Docker, but we would loose a lot of  isolation & stability.

samdark replied:
What is the problem here?

schmunk42 commented:
Example build: https://git.hrzg.de/schmunk42/yii2/pipelines/18450/builds

schmunk42 commented:
You can't really fake localhost in the dockerized setup.
It also makes no sense to build the database servers into the PHP image.

Can we just create another config file, which is loaded via ENV bar?

Von meinem iPhone gesendet

> Am 28.01.2017 um 11:58 schrieb Alexander Makarov <notifications@github.com>:
> 
> @samdark commented on this pull request.
> 
> In tests/data/config.php:
> 
> > @@ -16,29 +16,29 @@
>  $config = [
>      'databases' => [
>          'cubrid' => [
> -            'dsn' => 'cubrid:dbname=demodb;host=localhost;port=33000',
> +            'dsn' => 'cubrid:dbname=demodb;host=cubrid;port=33000',
> What is the problem here?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

samdark commented:
Yes, we can. Why not?

schmunk42 commented:
Was just a discussion :)
I'll add it to the PR...

Von meinem iPhone gesendet

> Am 28.01.2017 um 16:36 schrieb Alexander Makarov <notifications@github.com>:
> 
> Yes, we can. Why not?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

schmunk42 committed:
Merge commit '95b7ffc066c03f2510ef4e6681e1ddf1494e916f' into feature/docker-testing

# Conflicts:
#	tests/framework/db/sqlite/QueryBuilderTest.php

schmunk42 committed:
added separate docker config for testing

schmunk42 committed:
use custom docker config in dockerized tests

schmunk42 committed:
removed TODOs

schmunk42 commented:
Would be ready from my side.

SilverFire commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,118 @@
+  stage: test
+  script:
+    - docker-compose up --build -d
+    - docker-compose run --rm php vendor/bin/phpunit -v --exclude caching,db,data --log-junit tests/_junit/test.xml
```
`--rm` option is redundant container was started in background (`docker-compose up -d`). Or I am missing something?

cebe replied:
starting in the background is `-d`, removing the container after it has been stopped is `--rm`.

SilverFire replied:
The line I've commented has -rm argument, that according to the docs should remove container after the command execution, but ONLY when the container is NOT running in the background.

Line above (docker-compose up --build -d) stars container in background.
Does is make any sense to pass --rm flag for the RUN command in this case?

schmunk42 replied:
There are actually two containers running in line 30. The one in the background (webserver) and the actual run process which does the testing.
I'd say you usually (depends on the use-case) run `run` with `--rm`, I does not really matter in the CI, but it should be cleaner in general, see also https://github.com/docker/compose/issues/943

@mikehaertl @motin WDYT?

PS: Cleanup is actually done in `after_script`.

motin replied:
It is good practice and relevant in this case to clean up with "--rm" when using `run` since `run` always starts a new container which otherwise will be left behind on the Docker host.

SilverFire replied:
Yeah, I have mixed up `run` and `exec`. Thank you @schmunk42

SilverFire reviewed (changes requested):
Some bash files do not have a newline at the end.

Great job @schmunk42! Thank you!

samdark commented:
@SilverFire I don't have enough expertise to review it.

samdark commented:
@schmunk42 do we have to commit composer.lock?

cebe commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,118 @@
+    - docker-compose run --rm php vendor/bin/phpunit -v --group mssql
+    - docker-compose down -v --remove-orphans
+
+travis:
```
what does this have to do with travis?

schmunk42 replied:
Removed, thanks.

cebe commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM dmstr/php-yii2:7.0-fpm-1.9-beta2-alpine-nginx
```
is it possible to move the DockerFiles to some dir in tests/ ? if they need to be located in repo root I would prefer copying them from test folder here before starting docker.

schmunk42 replied:
This would imply that we use another way of getting the framework source-code onto the image, because a Dockerfile defines the build context and you can not add something from outside the build context.

An option is to use host-volumes, but they can problematic regarding performance and other things.
We could also clone or download the current commit when building an image, but that would be annoying in development.

I also don't like it, that these files are in the root folder, but it was the best option to me for now.

I am open for discussion.

cebe replied:
I was thinking about moving the Dockerfile in the build script, e.g. `cp tests/docker/Dockerfile-mysql Dockerfile` and then run docker-compose, so that they are not located in the repo root but are moved to repo root when needed.

schmunk42 replied:
Found a way :)

schmunk42 replied:
Actually, the one in the root folder could become kind of an official image, I am working with @mikehaertl on something similar.

Or should this be moved to tests also?

cebe replied:
I think it is fine like this now.

cebe reviewed (commented):
Looks great to me except the comments below. Good job, thank you!

cebe commented:
> @schmunk42 do we have to commit composer.lock?

composer.lock changes look good to me.

cebe commented:
> In tests, there are dependencies from i.e. log suite to db, which make tests fail which would actually 
> run on a vanilla PHP. I thought about also putting them into group db, WDYT?

Sure, all db related tests should be in the group db, that would also allow excluding them.
`phpunit --exclude=db` should run all tests that are not DB related.

schmunk42 committed:
cleaned cleanup (done in after_script)

schmunk42 committed:
added build trigger example

schmunk42 commented:
About composer.lock - https://getcomposer.org/doc/01-basic-usage.md#composer-lock-the-lock-file

You should update it from time to time and always for a release. A developer who wants to run `update` can still do so. You also need it for some CI related tasks, here and there.

schmunk42 committed:
added db group to test

schmunk42 committed:
Merge branch 'feature/docker-testing' into tests/mysql

schmunk42 committed:
added groups to tests

cebe commented on tests/framework/console/controllers/DbMessageControllerTest.php:
```diff
@@ -1,6 +1,10 @@
 namespace yiiunit\framework\console\controllers;
 
+/**
+ * @group db
```
this must be on the class phpdoc, not here.

schmunk42 committed:
Merge commit '22972b86dbcc6e010dc90df921090e713521f0b9' into feature/docker-testing

schmunk42 committed:
Merge branch 'tests/mysql' into feature/docker-testing

schmunk42 commented:
Current build state:
- tests/mysql https://gitlab.com/yiisoft-contrib/yii2/pipelines/6519636/builds
- tests/pgsql https://gitlab.com/yiisoft-contrib/yii2/pipelines/6519650/builds

schmunk42 commented:
Anything more?

samdark commented:
I think it's OK to merge. @cebe ?

schmunk42 committed:
moved Dockerfile(s) to test folders

schmunk42 committed:
updated changelog

cebe committed:
Merge branch 'master' into feature/docker-testing

cebe commented:
Merged, thank you for this great addition! :)

schmunk42 commented:
You're welcome :)

Auto-pull also works already, see: https://gitlab.com/yiisoft-contrib/yii2/pipelines/6851144
Do these tests also fail in Travis? I need to look that up.

I'll look after the tests from time to time. db, cache and mysql are green AFAIR.

You can push i.e. to `tests/pgsql` on gitlab.com to trigger the PostgreTest suite.

The "exotic" databases still need some testing of the testing...

samdark commented:
@schmunk42 is there a way to integrate these tests with GitHub? Some failures are about differences in intl/ICU data used. gitlab has newer version. Tests should be adjusted.

schmunk42 commented:
<img width="1208" alt="bildschirmfoto 2017-03-06 um 20 40 07" src="https://cloud.githubusercontent.com/assets/649031/23626420/229ab004-02ad-11e7-99f9-bf289f53b66c.png">

It's set up to sync hourly.

samdark commented:
Yes, I know. I've meant displaying links and results in pull requests etc.

schmunk42 commented:
Hmm, there's build badge, but further integration is limited AFAIR.
Related: https://gitlab.com/gitlab-org/gitlab-ce/issues/2595

schmunk42 commented:
> is there a way to integrate these tests with GitHub? Some failures are about differences in intl/ICU data used. gitlab has newer version. Tests should be adjusted.

@samdark You could run tests also on Travis, but we didn't want to touch the existing Travis setup. 
Looks like there's no support for multiple files: https://github.com/travis-ci/travis-ci/issues/3540

Could work with a `travis/docker` branch though.

samdark commented:
I see. I'd let travis alone as it is. It is sometimes weird but that gives additional cases to fix which is good.

schmunk42 commented:
OK, we'll be able to see these issues better, when we have the base images. Since we then can test multiple versions easily.

CC: @mikehaertl
